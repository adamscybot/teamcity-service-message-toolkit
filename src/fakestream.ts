import { Readable } from 'node:stream'
import { LinearStreamParser, TreeStreamParser } from './parser.js'

// Fake input stream
const inputStream = new Readable({
  read() {},
})

const linearStreamParser = new LinearStreamParser()
const treeStreamParser = new TreeStreamParser()

inputStream.pipe(linearStreamParser).pipe(treeStreamParser)

treeStreamParser.on('data', (message: any) => {
  console.log('A message was received:')
  console.log(message)
})

const logLines = `##teamcity[testSuiteStarted name='foo/__tests__/file.js' flowId='12345']
hsahldasd dsadas
dsadsa dsadas
dsadasdasd
##teamcity[testSuiteStarted name='path' flowId='12345']
##teamcity[testSuiteStarted name='to' flowId='12345']
##teamcity[testSuiteStarted name='test1' flowId='12345']
##teamcity[testStarted name='title1' flowId='12345']
##teamcity[testFailed name='title1' flowId='12345']
##teamcity[testFinished name='title1' duration='123' flowId='12345']
##teamcity[testSuiteFinished name='test1' flowId='12345']
##teamcity[testSuiteStarted name='test2' flowId='12345']
##teamcity[testStarted name='title2' flowId='12345']
##teamcity[testIgnored name='title2' message='pending' flowId='12345']
##teamcity[testFinished name='title2' duration='123' flowId='12345']
##teamcity[testSuiteFinished name='test2' flowId='12345']
##teamcity[testSuiteFinished name='to' flowId='12345']
##teamcity[testSuiteFinished name='path' flowId='12345']
##teamcity[testSuiteFinished name='foo/__tests__/file.js' flowId='12345']
##teamcity[testSuiteStarted name='foo/__tests__/file2.js' flowId='12345']
##teamcity[testSuiteStarted name='path2' flowId='12345']
##teamcity[testSuiteStarted name='to' flowId='12345']
##teamcity[testSuiteStarted name='test3' flowId='12345']
##teamcity[testStarted name='title3' flowId='12345']
##teamcity[testFinished name='title3' duration='123' flowId='12345']
##teamcity[testSuiteFinished name='test3' flowId='12345']
##teamcity[testSuiteStarted name='test4' flowId='12345']
##teamcity[testStarted name='title4' flowId='12345']
##teamcity[testFailed name='title5' message='Unexpected exception' details='at path/to/file1.js:1|n    at path/to/file2.js:2' flowId='12345']
##teamcity[testFinished name='title5' duration='123' flowId='12345']
##teamcity[testSuiteFinished name='test5' flowId='12345']
##teamcity[testSuiteStarted name='constructor' flowId='12345']
##teamcity[testStarted name='title6' flowId='12345']
##teamcity[testFinished name='title6' duration='123' flowId='12345']
##teamcity[testSuiteFinished name='constructor' flowId='12345']
##teamcity[testSuiteFinished name='to' flowId='12345']
##teamcity[testSuiteFinished name='path2' flowId='12345']
##teamcity[testSuiteFinished name='foo/__tests__/file2.js' flowId='12345']`

// Push lines into the stream with 1s interval
const lines = logLines.split('\n')
let index = 0
const intervalId = setInterval(() => {
  if (index < lines.length) {
    inputStream.push(lines[index] + '\n')
    index++
  } else {
    // All lines are pushed, close the stream
    inputStream.push(null)
    clearInterval(intervalId)
  }
}, 1000)
