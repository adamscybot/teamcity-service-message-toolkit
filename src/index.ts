export * from './messages'
export * from './tokeniser'
export * from './parser'
