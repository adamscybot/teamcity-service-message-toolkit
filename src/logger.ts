// @ts-nocheck
/**
 * Provides an API that allows you to configure a stream that simulates a test
 * runner that outputs TeamCity formatted log lines.
 */

import parseDuration from 'parse-duration'
import { Readable } from 'node:stream'
import { buildServiceMessage } from './parser/string-parse.js'

export type BeforeSuiteAddedCallback = (
  suite: TcDummySuite,
  index: number
) => void
export type BeforeTestAddedCallback = (
  suite: TcDummyTestIgnored | TcDummyTestSucceeds | TcDummyTestFails,
  index: number
) => void

abstract class TcStreamable {
  protected readableStream: Readable

  constructor(stream: Readable) {
    this.readableStream = stream
  }

  protected pushTcEvent(...args: Parameters<typeof buildServiceMessage>) {
    this.readableStream.push('\n' + buildServiceMessage(...args))
  }

  /**
   * Get access to the underlying `ReadableStream`. This can be used to listen
   * to events or manipulate the stream.
   *
   * @example
   *
   *     logger.addSuite((suite) => suite.name('Example suite')).stream((stream) => stream.)
   *
   * @param cb A callback which is passed the underlying `ReadableStream`
   */
  stream(cb: (stream: Readable) => void) {
    cb(this.readableStream)
    return this
  }

  abstract exec(): Promise<void>
}

abstract class TcContainsRawText extends TcStreamable {
  protected _stdOut: string = ''
  protected _stdErr: string = ''

  appendToStdOut(text: string) {
    this._stdOut += text
  }

  appendToStdErr(text: string) {
    this._stdErr += text
  }
}

export class TcDelay extends TcStreamable {
  private duration: string

  constructor(stream: Readable, duration: string) {
    super(stream)
    this.duration = duration
  }

  exec() {
    return new Promise<void>((resolve) => {
      setTimeout(() => {
        resolve()
      }, parseDuration(this.duration))
    })
  }
}

export abstract class TcBaseTest extends TcStreamable {
  protected nameString: string = 'Fake Test'

  name(name: string) {
    this.nameString = name
    return this
  }

  protected async pushTest(
    cb: () => void,
    opts?: { duration?: string; realTime?: boolean }
  ) {
    const { duration, realTime = false } = opts ?? {}
    this.pushTcEvent({
      messageName: 'testStarted',
      kwargs: {
        name: this.nameString,
      },
    })

    const durationMs = duration !== undefined ? parseDuration(duration) : null

    if (duration !== undefined && realTime) {
      const delayTask = new TcDelay(this.readableStream, duration)
      await delayTask.exec()
    }
    cb()

    this.pushTcEvent({
      messageName: 'testFinished',
      kwargs: {
        name: this.nameString,
        ...(typeof durationMs === 'number'
          ? { duration: `${durationMs}` }
          : {}),
      },
    })
  }
}

export abstract class TcDummyTestRunnable extends TcBaseTest {
  protected durationString: string | undefined = undefined
  protected realTime: boolean | undefined = undefined
  protected bodyContent: string | null = null

  duration(duration: string, opts?: { realTime?: boolean }) {
    this.durationString = duration
    this.realTime = opts?.realTime
    return this
  }

  body(content: string) {
    this.bodyContent = content
    return this
  }

  protected async pushRunnableTest(cb?: () => void) {
    await this.pushTest(
      () => {
        if (this.bodyContent !== null)
          this.readableStream.push(`\n${this.bodyContent}`)
        cb?.()
      },
      { duration: this.durationString, realTime: this.realTime }
    )
  }
}

export class TcDummyTestIgnored extends TcBaseTest {
  protected messageString: string = 'ignored'
  protected wrapped: boolean = false

  /**
   * Set the failure message.
   *
   * @param message The string containing the high level failure message
   */
  message(message: string) {
    this.messageString = message
    return this
  }

  /**
   * TC allows log output for ignored tests to be bookended by start/finish
   * markers or to be standalone. Setting this to true adds the bookending.
   * Useful for testing different parsing scenarios.
   */
  wrap(shouldWrap: boolean = true) {
    this.wrapped = shouldWrap
    return this
  }

  async exec() {
    const execIgnore = () =>
      this.pushTcEvent({
        messageName: 'testIgnored',
        kwargs: {
          name: this.nameString,
        },
      })

    if (this.wrapped) {
      await this.pushTest(() => execIgnore())
    } else {
      execIgnore()
    }
  }
}

export class TcDummyTestSucceeds extends TcDummyTestRunnable {
  async exec() {
    await this.pushRunnableTest()
  }
}

export class TcDummyTestFails extends TcDummyTestRunnable {
  private messageString: string | null = null
  private detailsString: string | null = null

  /**
   * Set the failure message.
   *
   * @param message The string containing the high level failure message
   */
  message(message: string) {
    this.messageString = message
    return this
  }

  /**
   * Set the details message, usually used for stack traces relating to the
   * error.
   *
   * @param message The string containing the additional error details
   */
  details(details: string) {
    this.detailsString = details
    return this
  }

  async exec() {
    await this.pushRunnableTest(() => {
      this.pushTcEvent({
        messageName: 'testFailed',
        kwargs: {
          ...(this.messageString !== null
            ? { message: this.messageString }
            : {}),
          ...(this.detailsString !== null
            ? { details: this.detailsString }
            : {}),
        },
      })
    })
  }
}

export class TcDummyTestAccessor {
  protected stream: Readable

  constructor(stream: Readable) {
    this.stream = stream
  }

  /**
   * Define the test as one which will been ignored.
   *
   * @returns A test configuration object for further customisation.
   */
  ignored() {
    return new TcDummyTestIgnored(this.stream)
  }

  /**
   * Define the test as one which will succeed.
   *
   * @returns A test configuration object for further customisation.
   */
  succeeds() {
    return new TcDummyTestSucceeds(this.stream)
  }

  /**
   * Define the test as one which will fail.
   *
   * @returns A test configuration object for further customisation.
   */
  fails() {
    return new TcDummyTestFails(this.stream)
  }
}

export class TcDummySuite extends TcContainsRawText {
  private beforeTestAddedCb: BeforeTestAddedCallback | undefined = undefined
  private testCounter: number = 1
  private nameString: string = ''
  private tasks: (
    | TcDummyTestIgnored
    | TcDummyTestSucceeds
    | TcDummyTestFails
    | TcDelay
  )[] = []

  constructor(
    stream: Readable,
    opts?: { beforeTestAdded?: BeforeTestAddedCallback }
  ) {
    super(stream)
    this.beforeTestAddedCb = opts?.beforeTestAdded
  }

  /**
   * Set the name of this suite.
   *
   * @example
   *
   *     logger.addSuite((suite) => suite.name('Example suite'))
   *
   * @param name The new name of the suite which will be shown in the log output
   */
  name(name?: string) {
    if (!name) return this.nameString
    this.nameString = name
    return this
  }

  /**
   * Add a an individual test to this suite and allow it to be configured.
   *
   * @example
   *
   *     logger.addSuite((suite) =>
   *       suite
   *         .name('Example suite')
   *         .addTest((test) => test.success().name('Example Test'))
   *     )
   *
   * @param cb A callback that is passed the new test for further configuration
   */
  addTest(
    cb: (
      suite: TcDummyTestAccessor
    ) => TcDummyTestIgnored | TcDummyTestSucceeds | TcDummyTestFails
  ) {
    const newTest = cb(new TcDummyTestAccessor(this.readableStream))

    if (this.beforeTestAddedCb) {
      this.beforeTestAddedCb(newTest, this.testCounter)
    }

    this.tasks.push(newTest)
    this.testCounter++
    return this
  }

  /**
   * Add a delay between the output of each test
   *
   * @example
   *
   *     logger.addSuite((suite) =>
   *       suite
   *         .name('Example suite')
   *         .addTest((test) => test.success().name('Example Test'))
   *         .delay('10s')
   *         .addTest((test) => test.success().name('Delayed Test'))
   *     )
   *
   * @param duration A duration in string form compatible with the
   *   `parse-duration` duration library. E.g. 10s
   */
  delay(duration: string) {
    this.tasks.push(new TcDelay(this.readableStream, duration))
    return this
  }

  async exec() {
    this.pushTcEvent({
      messageName: 'testSuiteStarted',
      kwargs: { name: this.nameString },
    })

    for (let task of this.tasks) {
      await task.exec()
    }

    this.pushTcEvent({
      messageName: 'testSuiteFinished',
      kwargs: { name: this.nameString },
    })
  }
}

export class TcDummyTestLogger extends TcStreamable {
  private suites: TcDummySuite[] = []
  private suiteCounter: number = 1
  private beforeSuiteAddedCb: BeforeSuiteAddedCallback | undefined = undefined
  private beforeTestAddedCb: BeforeTestAddedCallback | undefined = undefined

  /**
   * Add a test suite and allow it to be configured.
   *
   * @example
   *
   *     logger.addSuite((suite) => suite.name('Example suite'))
   *
   * @param cb A callback that is passed the new suite for further configuration
   */
  addSuite(cb: (suite: TcDummySuite) => TcDummySuite) {
    const newSuite = cb(
      new TcDummySuite(this.readableStream, {
        beforeTestAdded: this.beforeTestAddedCb,
      })
    )

    if (this.beforeSuiteAddedCb) {
      this.beforeSuiteAddedCb(newSuite, this.suiteCounter)
    }

    this.suites.push(newSuite)
    this.suiteCounter++
    return this
  }

  addOrReplaceSuite(name: string, cb: (suite: TcDummySuite) => TcDummySuite) {
    const existingSuite = this.suites.find((suite) => suite.name() === name)

    const newSuite = cb(
      existingSuite ??
        new TcDummySuite(this.readableStream, {
          beforeTestAdded: this.beforeTestAddedCb,
        })
    )

    if (this.beforeSuiteAddedCb) {
      this.beforeSuiteAddedCb(newSuite, this.suiteCounter)
    }

    this.suites.push(newSuite)
    this.suiteCounter++
    return this
  }

  beforeSuiteAdded(cb: BeforeSuiteAddedCallback) {
    this.beforeSuiteAddedCb = cb
    return this
  }

  beforeTestAdded(cb: BeforeTestAddedCallback) {
    this.beforeTestAddedCb = cb
    return this
  }
  /**
   * Start outputting the configured logging to the stream.
   *
   * @returns {Promise} A promise that resolves once all of the requested log
   *   output is outputted to the stream.
   */
  async exec() {
    for (const suite of this.suites) {
      await suite.exec()
    }
    this.readableStream.push(null)
  }

  /**
   * Configure the stream to pipe into another stream. This is provided as a
   * utility wrapper around `logger.stream(stream => stream.pipe()`
   *
   * @example Output all the logs to stdout
   *
   *     logger.pipe(process.stdout)
   */
  pipe(...args: Parameters<typeof this.readableStream.pipe>) {
    this.readableStream.pipe(...args)
    return this
  }

  processMessage() {}
}

export default () =>
  new TcDummyTestLogger(new Readable({ encoding: 'utf8', read: () => {} }))
