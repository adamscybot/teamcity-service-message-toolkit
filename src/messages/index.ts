export * from './builder'
export * from './constants'
export * from './repository'
export * from './schema'
export * from './types'
export * from './utils'
