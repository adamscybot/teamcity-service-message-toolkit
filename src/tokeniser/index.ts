import { TokeniserStream } from './stream'
import { TokeniserOpts } from './lexer'
export { TokenTypes } from './lexer'

export type * from './stream'
export type * from './lexer'

export default {
  stream(opts?: TokeniserOpts) {
    return new TokeniserStream()
  },
}
