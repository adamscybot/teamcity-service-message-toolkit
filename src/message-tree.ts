import { Readable } from 'node:stream'
import messages, {
  TcTestStarted,
  TcTestFinished,
  TcTestStdErr,
  TcTestStdOut,
  TcTestIgnored,
  TcTestFailed,
  TcTestSuiteFinished,
  TcTestSuiteStarted,
  Messages,
} from './messages/wrappers/default-messages.js'
import { TcBaseMessage } from './messages/wrappers/base-message-types.js'

type InstanceTypeTuple<T extends Array<(typeof messages)[number]>> = {
  [P in keyof T]: InstanceType<T[P]>
}

class TcTreeNode<
  AllowedMessages extends (typeof messages)[number],
  RequiredMessages extends Array<AllowedMessages> = []
> {
  public messages: [
    ...InstanceTypeTuple<RequiredMessages>,
    ...Array<InstanceType<AllowedMessages>>
  ]

  constructor(requiredMessages: InstanceTypeTuple<RequiredMessages>) {
    this.messages = requiredMessages as [
      ...InstanceTypeTuple<RequiredMessages>,
      ...Array<InstanceType<AllowedMessages>>
    ]
  }

  appendMessage(message: InstanceType<AllowedMessages>) {
    this.messages.push(message)
  }

  public findMessage<
    MessageName extends RequiredMessages[number]['messageName']
  >(
    messageName: MessageName
  ): InstanceType<
    Extract<RequiredMessages[number], { messageName: MessageName }>
  >
  public findMessage<MessageName extends AllowedMessages['messageName']>(
    messageName: MessageName
  ):
    | InstanceType<Extract<AllowedMessages, { messageName: MessageName }>>
    | undefined
  public findMessage<
    MessageName extends
      | AllowedMessages['messageName']
      | RequiredMessages[number]['messageName']
  >(
    messageName: MessageName
  ):
    | InstanceType<
        Extract<RequiredMessages[number], { messageName: MessageName }>
      >
    | InstanceType<Extract<AllowedMessages, { messageName: MessageName }>>
    | undefined {
    return this.messages.find(
      (message) => message.messageName() === messageName
    )
  }
}

abstract class TcContainsRawText<
  AllowedMessages extends (typeof messages)[number],
  RequiredMessages extends Array<AllowedMessages>
> extends TcTreeNode<AllowedMessages, RequiredMessages> {
  protected _stdOut: string = ''
  protected _stdErr: string = ''

  appendToStdOut(text: string) {
    this._stdOut += text
  }

  appendToStdErr(text: string) {
    this._stdErr += text
  }
}

export abstract class TcBaseTest<
  AllowedMessages extends (typeof messages)[number],
  RequiredMessages extends Array<AllowedMessages> = []
> extends TcContainsRawText<
  | typeof TcTestStarted
  | typeof TcTestFinished
  | typeof TcTestStdOut
  | typeof TcTestStdErr
  | AllowedMessages,
  RequiredMessages
> {
  protected nameString: string

  name(name: string) {
    this.nameString = name
    return this
  }
}

export abstract class TcDummyTestRunnable<
  AllowedMessages extends (typeof messages)[number],
  RequiredMessages extends Array<AllowedMessages> = []
> extends TcBaseTest<typeof TcTestStarted, [typeof TcTestStarted]> {
  protected durationString: string | undefined = undefined

  duration(duration: string) {
    const test = this.messages.find(
      (message) => message instanceof TcTestStarted
    )
    this.durationString = duration
    const test = this.messages
    return this
  }
}

export class TcDummyTestIgnored extends TcBaseTest<
  typeof TcTestIgnored,
  [typeof TcTestIgnored]
> {
  protected messageString: string

  /**
   * Set the failure message.
   *
   * @param message The string containing the high level failure message
   */
  message(message: string) {
    this.messageString = message
    return this
  }

  /**
   * TC allows log output for ignored tests to be bookended by start/finish markers or to be standalone.
   * Setting this to true adds the bookending. Useful for testing different parsing scenarios.
   **/
  wrap(shouldWrap: boolean = true) {
    this.wrapped = shouldWrap
    return this
  }
}

export class TcDummyTestSucceeds extends TcDummyTestRunnable<never, never> {}

export class TcDummyTestFails extends TcDummyTestRunnable<
  typeof TcTestFailed,
  [typeof TcTestFailed]
> {
  private messageString: string | null = null
  private detailsString: string | null = null

  /**
   * Set the failure message.
   *
   * @param message The string containing the high level failure message
   */
  message(message: string) {
    this.messageString = message
    return this
  }

  /**
   * Set the details message, usually used for stack traces relating to the error.
   *
   * @param message The string containing the additional error details
   */
  details(details: string) {
    this.detailsString = details
    return this
  }
}

export class TcDummyTestAccessor {
  protected stream: Readable

  constructor(stream: Readable) {
    this.stream = stream
  }

  /**
   * Define the test as one which will been ignored.
   *
   * @returns a test configuration object for further customisation.
   */
  ignored() {
    return new TcDummyTestIgnored()
  }

  /**
   * Define the test as one which will succeed.
   *
   * @returns a test configuration object for further customisation.
   */
  succeeds() {
    return new TcDummyTestSucceeds()
  }

  /**
   * Define the test as one which will fail.
   *
   * @returns a test configuration object for further customisation.
   */
  fails() {
    return new TcDummyTestFails()
  }
}

export class TcSuiteNode extends TcContainsRawText<
  typeof TcTestSuiteStarted | typeof TcTestSuiteFinished,
  [typeof TcTestSuiteStarted]
> {
  private childSuites: TcSuiteNode[] = []

  getChildSuites() {
    return this.childSuites
  }

  addChildSuite(child: TcSuiteNode) {
    this.childSuites.push(child)
    return this
  }

  getName() {
    return this.findMessage('testSuiteStarted').getName()
  }
}

// Do flows inherit the parent context? What are the semantics there?
// I think when a flow is seen for first time, its context is decided at that point.
// It has to be, otherwis eyou couldnt nest parrallelised tests.
//aCTUALLY IT seems suite start/stop and test start/stop calls have to be on the same flows
export class TcMessageFlow {
  private subFlows: TcMessageFlow[] = []
  private parent: TcMessageFlow | undefined = undefined
  // this wont work, we need to persist a tree
  // I ahve other problems. Like some messages being able to be children of suites and tests.
  // but mayube thats not really a problem as i can just have them as members where they make sense only
  // but wait. flows aren't supposed to last forever and can be closed. so anything owned here would die.
  // maybe this should just find the right thing to append to.
  // what if i passed the tree route into the flow register and then into here?
  // i probs do need a proper tree structure with pruning thoguh
  // it would be nice also to be able to see all the messages in a flow. filte rby flowid. but this would be hard to get in the right order, unless i store timestamps
  // of message arrival.
  // i think flow should really store the messages as this is supposed ot be lgon lived strucute
  private contextStack: Array<TcContainsRawText<any, any>> = []

  constructor(parent?: TcMessageFlow) {
    this.parent = parent
  }

  private peekContext() {
    if (this.contextStack.length < 1) return undefined
    return this.contextStack[this.contextStack.length - 1]
  }

  private pushContext(unit: TcContainsRawText<any, any>) {
    this.contextStack.push(unit)
  }

  public addSubFlow() {
    const subFlow = new TcMessageFlow(this)
    this.subFlows.push(subFlow)
    return subFlow
  }

  private processTestSuiteStarted(message: TcTestSuiteStarted) {
    const contextHead = this.peekContext()
    if (!(contextHead instanceof TcSuiteNode) && contextHead !== undefined) {
      throw new Error(
        `Attempted to parse message '${message.messageName()}' but the context head is ${typeof contextHead}. Suites can only be started inside other suites or at the top level of a flow.`
      )
    }

    const suite = new TcSuiteNode([message])
    this.pushContext(suite)
    return suite
  }

  public processMessage(message: InstanceType<Messages>) {
    if (message instanceof TcTestSuiteStarted) {
      const suite = new TcSuiteNode([message])
      this.pushContext(suite)
      return
    }

    if (message instanceof TcTestStarted) {
      const name = message.getName()
      if (
        this.currentLogContext === undefined ||
        !(this.currentLogContext instanceof TcSuiteNode)
      ) {
        console.error(
          `Message of type '${message.messageName}' received whilst not in the context of a containing suite.`
        )
        return
      }

      this.currentLogContext.appendMessage(message)
      return
    }

    if (message instanceof TcTestSuiteFinished) {
      this.currentLogContext = undefined
      return
    }

    console.warn(`Message "${message.messageName}" not currently supported.`)
  }
}

export class TcFlowRegistry {
  private static rootFlowId = Symbol('rootFlow')
  private flowMap = new Map<
    string | typeof TcFlowRegistry.rootFlowId,
    TcMessageFlow
  >()
  private activeFlowId: string | typeof TcFlowRegistry.rootFlowId =
    TcFlowRegistry.rootFlowId

  constructor() {
    const rootFlow = new TcMessageFlow()
    this.flowMap.set(TcFlowRegistry.rootFlowId, rootFlow)
  }

  processMessage(message: InstanceType<Messages>) {
    const flowId = message.flowId()

    let flowToPushTo = this.getRootFlow()
    if (flowId) {
      const addressedFlow = this.flowMap.get(flowId)
      if (!addressedFlow) {
        // Is this where the context is copied?
        this.flowMap.set(flowId, this.getActiveFlow().addSubFlow())
      }
      this.activeFlowId = flowId
      flowToPushTo = this.getActiveFlow()
    }

    flowToPushTo.processMessage(message)

    return flowToPushTo
  }

  getFlow(flowId: string | typeof TcFlowRegistry.rootFlowId) {
    return this.flowMap.get(flowId)
  }

  getRootFlow() {
    return this.getFlow(TcFlowRegistry.rootFlowId)!
  }

  getActiveFlow() {
    return this.getFlow(this.activeFlowId)!
  }
}

export class TcMessageTree {
  private suites: TcSuiteNode[] = []
  private
  private currentLogContext: Array<TcContainsRawText<any, any>>

  public findSuiteByName(name: string) {
    return this.suites.find((suite) => suite.getName() === name)
  }

  public getSuites() {
    return this.suites
  }
}

export default () => new TcMessageTree()
