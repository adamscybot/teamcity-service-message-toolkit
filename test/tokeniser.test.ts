import { describe, expect } from 'vitest'

import { test } from './fixtures/wrappers'
import { TokenTypes, TokeniserOpts, tokeniser } from '../src/tokeniser'
import { Token } from 'moo'

const getAllTokens = (line: string, opts?: TokeniserOpts) => {
  const lexer = tokeniser(opts)
  lexer.reset(line)
  return Array.from(lexer)
}

describe('Tokeniser', () => {
  describe('Lexer foundation', () => {
    test('Tokenises single attribute message', () => {
      expect(
        getAllTokens("##teamcity[testMessage attributeExample='hello']")
      ).toMatchSnapshot()
    })

    test('Escapes string literals', () => {
      expect(
        getAllTokens(
          "##teamcity[testMessage attributeExample='|[Diff|cult|r|nmes|'sage |0x0134|]']"
        )
      ).toMatchSnapshot()
    })
  })
})
