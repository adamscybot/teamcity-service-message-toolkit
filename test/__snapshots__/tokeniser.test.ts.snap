// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Tokeniser > Lexer foundation > Escapes string literals 1`] = `
[
  {
    "col": 1,
    "line": 1,
    "lineBreaks": 0,
    "offset": 0,
    "text": "##teamcity",
    "toString": [Function],
    "type": "serviceMessageIdent",
    "value": "##teamcity",
  },
  {
    "col": 11,
    "line": 1,
    "lineBreaks": 0,
    "offset": 10,
    "text": "[",
    "toString": [Function],
    "type": "parametersOpen",
    "value": "[",
  },
  {
    "col": 12,
    "line": 1,
    "lineBreaks": 0,
    "offset": 11,
    "text": "testMessage",
    "toString": [Function],
    "type": "messageName",
    "value": "testMessage",
  },
  {
    "col": 23,
    "line": 1,
    "lineBreaks": 0,
    "offset": 22,
    "text": " ",
    "toString": [Function],
    "type": "space",
    "value": " ",
  },
  {
    "col": 24,
    "line": 1,
    "lineBreaks": 0,
    "offset": 23,
    "text": "attributeExample",
    "toString": [Function],
    "type": "propertyName",
    "value": "attributeExample",
  },
  {
    "col": 40,
    "line": 1,
    "lineBreaks": 0,
    "offset": 39,
    "text": "=",
    "toString": [Function],
    "type": "propertyAssignOperator",
    "value": "=",
  },
  {
    "col": 41,
    "line": 1,
    "lineBreaks": 0,
    "offset": 40,
    "text": "'",
    "toString": [Function],
    "type": "stringStart",
    "value": "'",
  },
  {
    "col": 42,
    "line": 1,
    "lineBreaks": 0,
    "offset": 41,
    "text": "|[",
    "toString": [Function],
    "type": "stringEscapeSingleChar",
    "value": "[",
  },
  {
    "col": 44,
    "line": 1,
    "lineBreaks": 0,
    "offset": 43,
    "text": "Diff",
    "toString": [Function],
    "type": "stringValue",
    "value": "Diff",
  },
  {
    "col": 48,
    "line": 1,
    "lineBreaks": 0,
    "offset": 47,
    "text": "|c",
    "toString": [Function],
    "type": "stringEscapeSingleChar",
    "value": "c",
  },
  {
    "col": 50,
    "line": 1,
    "lineBreaks": 0,
    "offset": 49,
    "text": "ult",
    "toString": [Function],
    "type": "stringValue",
    "value": "ult",
  },
  {
    "col": 53,
    "line": 1,
    "lineBreaks": 0,
    "offset": 52,
    "text": "|r",
    "toString": [Function],
    "type": "stringEscapeSingleChar",
    "value": "
",
  },
  {
    "col": 55,
    "line": 1,
    "lineBreaks": 0,
    "offset": 54,
    "text": "|n",
    "toString": [Function],
    "type": "stringEscapeSingleChar",
    "value": "
",
  },
  {
    "col": 57,
    "line": 1,
    "lineBreaks": 0,
    "offset": 56,
    "text": "mes",
    "toString": [Function],
    "type": "stringValue",
    "value": "mes",
  },
  {
    "col": 60,
    "line": 1,
    "lineBreaks": 0,
    "offset": 59,
    "text": "|'",
    "toString": [Function],
    "type": "stringEscapeSingleChar",
    "value": "'",
  },
  {
    "col": 62,
    "line": 1,
    "lineBreaks": 0,
    "offset": 61,
    "text": "sage ",
    "toString": [Function],
    "type": "stringValue",
    "value": "sage ",
  },
  {
    "col": 67,
    "line": 1,
    "lineBreaks": 0,
    "offset": 66,
    "text": "|0x0134",
    "toString": [Function],
    "type": "stringEscapeUnicode",
    "value": "Ä´",
  },
  {
    "col": 74,
    "line": 1,
    "lineBreaks": 0,
    "offset": 73,
    "text": "|]",
    "toString": [Function],
    "type": "stringEscapeSingleChar",
    "value": "]",
  },
  {
    "col": 76,
    "line": 1,
    "lineBreaks": 0,
    "offset": 75,
    "text": "'",
    "toString": [Function],
    "type": "stringEnd",
    "value": "'",
  },
  {
    "col": 77,
    "line": 1,
    "lineBreaks": 0,
    "offset": 76,
    "text": "]",
    "toString": [Function],
    "type": "parametersClose",
    "value": "]",
  },
]
`;

exports[`Tokeniser > Lexer foundation > Tokenises single attribute message 1`] = `
[
  {
    "col": 1,
    "line": 1,
    "lineBreaks": 0,
    "offset": 0,
    "text": "##teamcity",
    "toString": [Function],
    "type": "serviceMessageIdent",
    "value": "##teamcity",
  },
  {
    "col": 11,
    "line": 1,
    "lineBreaks": 0,
    "offset": 10,
    "text": "[",
    "toString": [Function],
    "type": "parametersOpen",
    "value": "[",
  },
  {
    "col": 12,
    "line": 1,
    "lineBreaks": 0,
    "offset": 11,
    "text": "testMessage",
    "toString": [Function],
    "type": "messageName",
    "value": "testMessage",
  },
  {
    "col": 23,
    "line": 1,
    "lineBreaks": 0,
    "offset": 22,
    "text": " ",
    "toString": [Function],
    "type": "space",
    "value": " ",
  },
  {
    "col": 24,
    "line": 1,
    "lineBreaks": 0,
    "offset": 23,
    "text": "attributeExample",
    "toString": [Function],
    "type": "propertyName",
    "value": "attributeExample",
  },
  {
    "col": 40,
    "line": 1,
    "lineBreaks": 0,
    "offset": 39,
    "text": "=",
    "toString": [Function],
    "type": "propertyAssignOperator",
    "value": "=",
  },
  {
    "col": 41,
    "line": 1,
    "lineBreaks": 0,
    "offset": 40,
    "text": "'",
    "toString": [Function],
    "type": "stringStart",
    "value": "'",
  },
  {
    "col": 42,
    "line": 1,
    "lineBreaks": 0,
    "offset": 41,
    "text": "hello",
    "toString": [Function],
    "type": "stringValue",
    "value": "hello",
  },
  {
    "col": 47,
    "line": 1,
    "lineBreaks": 0,
    "offset": 46,
    "text": "'",
    "toString": [Function],
    "type": "stringEnd",
    "value": "'",
  },
  {
    "col": 48,
    "line": 1,
    "lineBreaks": 0,
    "offset": 47,
    "text": "]",
    "toString": [Function],
    "type": "parametersClose",
    "value": "]",
  },
]
`;
